### ISSUE TITLE
Header Missing - ESI Block Caching

### ISSUE THREAD

**Issue by JuanjoR** [nexcess/magento-turpentine#1210]
I have a problem activating ESI Block Caching "header"

Magento versión 1.9.2.2, Varnish 3.0.5 rev 1a89b1f Turpentine 0.7.0 I have -p esi_syntax=0x3 in DAEMONT_OPTS

Only caching "header", the header of the page disappears completely
> 		<reference name="header">
            <action method="setEsiOptions">
                <params>
                    <access>private</access>
                    <flush_events>
                        <wishlist_item_save_after/>
                        <wishlist_item_delete_after/>
                        <sales_quote_save_after/>
						<controller_action_layout_render_before_catalogsearch_result_index />
                    </flush_events>
                </params>
            </action>
        </reference>

In the turpentine.log

> 2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 1
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Matched router: standard
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Matched router: standard
2016-06-19T01:56:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_messages
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 0
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Matched router: standard
2016-06-19T01:56:45+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: messages
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 0
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Generated form_key: gJl69L6UzLkfm9vw
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 0
2016-06-19T01:56:45+00:00 DEBUG (7): TURPENTINE: Matched router: standard

This is the "header" of my theme local.xml

>         <reference name="header">
            <block type="catalog/product_compare_list" template="catalog/product/compare/list.phtml"/>
            <block type="page/switch" name="store_language" as="store_language" template="page/switch/flags.phtml"/>
            <block type="checkout/cart_sidebar" name="cart_sidebar" template="checkout/cart/sidebar.phtml" before="-">
                <action method="addItemRender"><type>simple</type><block>checkout/cart_item_renderer</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>grouped</type><block>checkout/cart_item_renderer_grouped</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>configurable</type><block>checkout/cart_item_renderer_configurable</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <block type="core/text_list" name="cart_sidebar.extra_actions" as="extra_actions" translate="label" module="checkout">
                    <label>Shopping Cart Sidebar Extra Actions</label>
                </block>
            </block>
            <block type="catalog/product_compare_sidebar" name="catalog.compare.header" as="header_compare" template="catalog/product/compare/header.phtml"/>
            <block type="wishlist/customer_sidebar" name="wishlist.header" as="header_wishlist" template="wishlist/header.phtml"/>
            <block type="page/template_links" name="top.links1" as="topLinks1" template="page/html/top.links.phtml"/>
            <action ifconfig="themesettings/header/displayheader_wishlistbox" method="unsetChild"><name>header_wishlist</name></action>
            <action ifconfig="themesettings/header/displayheader_comparebox" method="unsetChild"><name>header_compare</name></action>
        </reference>

And this is the checkout.xml 

>     <default>
        
        <reference name="top.links">
            <block type="checkout/links" name="checkout_cart_link">
                <action method="addCartLink"></action>
                <action method="addCheckoutLink"></action>
            </block>
        </reference>
        <reference name="right">
            <block type="checkout/cart_sidebar" name="cart_sidebar" template="checkout/cart/sidebar.phtml" before="-">
                <action method="addItemRender"><type>simple</type><block>checkout/cart_item_renderer</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>grouped</type><block>checkout/cart_item_renderer_grouped</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>configurable</type><block>checkout/cart_item_renderer_configurable</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <block type="core/text_list" name="cart_sidebar.extra_actions" as="extra_actions" translate="label" module="checkout">
                    <label>Shopping Cart Sidebar Extra Actions</label>
                </block>
            </block>
        </reference>
    </default>

After a week doing tests, I get not avoid two things

1) Prevent caching in header
2) Prevent caching cart_sidebar and top.link

I think the first thing you need, is to get the header caching, without that disappears the head of the page

I would appreciate help and excused for my bad English

**Comment by mobweb** [comment]
I actually have the same problem. Does this happen to you on every page load, or only some times? Because for me it only happens some times, and I have not been able to find a pattern yet.

**Comment by JuanjoR** [comment]
To me always happens, except, obviously, the not cached pages /custormer /checkout etc

You can see the cached page here

[http://www.thehouseofcarnival.com/](http://www.thehouseofcarnival.com/)

And not cached page (showing the header) here

[http://www.thehouseofcarnival.com/checkout/cart/](http://www.thehouseofcarnival.com/checkout/cart/)

**Comment by aricwatson** [comment]
Check your turpentine log to see if there's a message that says "Checking ESI block candidate: messages". Do you see the html comments before and after where the header should be? It looks to me like turpentine is not inserting an ESI tag where your header is.

Are there any errors in the main system log?

**Comment by JuanjoR** [comment]
The problem was in the magicmegamenu module, only to turn it off, everything works perfectly

However, **I have another theme with the same problem**. This second theme display the menu, but when active <reference name = "header">, the header disappears

Magento versión 1.9.2.0, Varnish 3.0.5 rev 1a89b1f Turpentine 0.7.0 I have -p esi_syntax=0x3 in DAEMONT_OPTS

>         <reference name="header">
            <action method="setEsiOptions">
                <params>
                    <access>private</access>
                    <flush_events>
                        <wishlist_item_save_after/>
                        <wishlist_item_delete_after/>
                        <sales_quote_save_after/>
						<controller_action_layout_render_before_catalogsearch_result_index />
                    </flush_events>
                </params>
            </action>
        </reference> 

In the turpentine.log
> 2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: footer
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: before_body_end
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_cookie_notice
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Block check passed, injecting block: global_cookie_notice
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_back_totop
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_popup_content_wrapper
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_ajax_addto
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: ajax.cart
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: luka.google_conversion_tracking
2016-06-20T16:54:12+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: core_profiler
2016-06-20T16:54:12+00:00 DEBUG (7): TURPENTINE: Set Varnish cache flag header to: 1
2016-06-20T16:54:12+00:00 DEBUG (7): TURPENTINE: Set ESI flag header to: 1
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: root
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: head
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: js_cookies
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_variation_css
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: optional_zip_countries
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: google_analytics
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_ajaxcart_head
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_multlidealpro_head
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: labels.css
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: top.Js
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: after_body_start
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: cookiealert
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: turpentine_notices
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: global_notices
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: header
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: store_mobile_language
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: mobile_currency
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em_header_mobile_top_links
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: em0132.top.menu.mobile
2016-06-20T16:54:25+00:00 INFO (6): TURPENTINE: Checking ESI block candidate: fblock_layout_handle_em0132-mobile-menu

this is the checkout.xml

>     <default>
        
        
        
        <reference name="header">
            <block type="checkout/cart_sidebar" name="cart_sidebar" template="checkout/cart/sidebar.phtml">
                <action method="addItemRender"><type>simple</type><block>checkout/cart_item_renderer</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>grouped</type><block>checkout/cart_item_renderer_grouped</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <action method="addItemRender"><type>configurable</type><block>checkout/cart_item_renderer_configurable</block><template>checkout/cart/sidebar/default.phtml</template></action>
                <block type="core/text_list" name="cart_sidebar.extra_actions" as="extra_actions" translate="label" module="checkout">
                    <label>Shopping Cart Sidebar Extra Actions</label>
                </block>
            </block>
        </reference>
    </default>

Here you can see a cached page:
[http://www.lamaisonducarnaval.be/](http://www.lamaisonducarnaval.be/)
The login page is not cached, you can see the header
[http://www.lamaisonducarnaval.be/customer/account/login/](http://www.lamaisonducarnaval.be/customer/account/login/)

I disabled from the administrator, all modules unrelated magento, but still does not show the header

**Comment by JuanjoR** [comment]
@aricwatson system.log shows the same information as the turpentine.log, no additional errors

**Comment by JuanjoR** [comment]
Solved. He had a module to minimize the html and the module that prevented from entering the code ESI

**Comment by aricwatson** [comment]
@JuanjoR Nice! Good find :)

**Comment by mobweb** [comment]
My issue is probably quite different, as it does not happen on every page load. I will investigate further and maybe open up a separate issue.

**Comment by pieterjanliekens** [comment]
Did you ever find the solution to your problem? @mobweb

**Comment by mobweb** [comment]
sorry @pieterjanliekens, I do not recall how I solved this.

**Comment by pieterjanliekens** [comment]
Can you check?
